i2c/imx6ull/libimx6ull-i2c.c

```diff
@@ -0,0 +1,361 @@
+		pctl.iomux.mode = i2c_pctl_mux[i2c.dev_no - 1][i].val;
+		platformctl(&pctl);
+	}
+
+	/* set IOsel */
+	pctl.type = pctl_ioisel;
+
+	for (i = 0; i < 2; ++i) {
+		pctl.ioisel.isel = i2c_pctl_isel[i2c.dev_no - 1][i].pctl;
+		pctl.ioisel.daisy = i2c_pctl_isel[i2c.dev_no - 1][i].val;
+		platformctl(&pctl);
+	}
+
+	/* PAD configuration taken from linux Device Tree, not needed for our current targets, leaving commented-out
+	 * as it might need to be enabled for targets with missing pullup resistors (and written in generic way) */
+#if 0
+	pctl.type = pctl_iopad;
+	pctl.iopad.pad = pctl_pad_gpio1_02;
+	pctl.iopad.hys = 1;
+	pctl.iopad.pus = 0b10; /* 100kOhm pull-up */
+	pctl.iopad.pue = 0;    /* pull/keeper enabled */
+	pctl.iopad.pke = 1;
+	pctl.iopad.ode = 1;      /* open drain enable */
+	pctl.iopad.speed = 0b10; /* speed - 100 MHz */
+	pctl.iopad.dse = 0x4;
+	pctl.iopad.sre = 0;
+
+	platformctl(&pctl);
+
+	pctl.iopad.pad = pctl_pad_gpio1_03;
+	platformctl(&pctl);
+#endif
+}
+
+
+static int is_bus_busy(void)
+{
+	return *(i2c.base + i2sr) & (1 << 5);
+}
+
+
+static int wait_bus_busy(int for_busy)
+{
+	int timeout = 10; /* x 10 us = 100us */
+
+	while (!!is_bus_busy() != !!for_busy) {
+		usleep(10);
+		if (--timeout <= 0)
+			return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+
+static int i2c_intr(unsigned int intr, void *data)
+{
+	/* clear interrupt */
+	*(i2c.base + i2sr) &= ~(1 << 1);
+
+	return i2c.cond;
+}
+
+
+static int i2c_trx_complete(void)
+{
+	const time_t timeout_us = 100; /* 100 us */
+	int ret = 0;
+
+	mutexLock(i2c.lock);
+
+	/* wait for transfer complete interrupt - with timeout */
+	while (!(*(i2c.base + i2sr) & (1 << 7))) {
+		if ((ret = condWait(i2c.cond, i2c.lock, timeout_us)) == -ETIME) {
+			ret = -ETIMEDOUT;
+			break;
+		}
+	}
+
+	/* clear interrupt second time - sticky conditionals */
+	*(i2c.base + i2sr) &= ~(1 << 1);
+
+	mutexUnlock(i2c.lock);
+	return ret;
+}
+
+
+static int i2c_acked(void)
+{
+	if (*(i2c.base + i2sr) & (1 << 0))
+		return -EIO;
+
+	return 0;
+}
+
+
+/* writes 1 byte and waits for completion / ack */
+static int write_byte(uint8_t byte)
+{
+	int ret = 0;
+
+	*(i2c.base + i2sr) = 0; /* clear previous interrupts */
+	*(i2c.base + i2dr) = byte;
+
+	/* wait for byte to be sent */
+	ret = i2c_trx_complete();
+	if (ret < 0)
+		return ret;
+
+	ret = i2c_acked();
+	return ret;
+}
+
+
+/* Performs i2c generic write operation to the given slave device. */
+int i2c_busWrite(uint8_t dev_addr, const uint8_t *data, uint32_t len)
+{
+	unsigned int i;
+	int ret;
+
+	/* single-master mode, bus should always be idle when we're not transferring */
+	if (is_bus_busy())
+		return -EBUSY;
+
+	/* generate START: enable master mode, enable TX, enable interrupt */
+	*(i2c.base + i2cr) |= (1 << 6) | (1 << 5) | (1 << 4);
+
+	/* wait for the bus to be busy */
+	ret = wait_bus_busy(1);
+
+	/* send first byte - DEVICE WRITE request */
+	if (ret == 0)
+		ret = write_byte((dev_addr << 1) | 0x0);
```

Comment:
`| 0x0` does nothing
```suggestion
		ret = write_byte(dev_addr << 1);
```

Reply:
I know the code is a no-op and will be optimized out but it highlights the intent (it's symmetric to `i2c_busRead`). 

I can define `I2C_OPERATION_WRITE` and `I2C_OPERATION_READ` if that is more preferable.